File: Algorithm/quicksort/quicksort.cpp
#include <iostream>
#include <utility>
#include <vector>

int partition(auto& nums, int low, int high) {
    int pivot = nums[high];
    int l = low;
    for (int r = low; r <= high - 1; r++) {
        if (nums[r] <= pivot) {
            std::swap(nums[l], nums[r]);
            l++;
        }
    }
    std::swap(nums[l], nums[high]);
    return l;
}

void helper(auto& nums, int low, int high) {
    if (low >= high) {
        return;
    }
    int pivot = partition(nums, low, high);
    helper(nums, low, pivot - 1);
    helper(nums, pivot + 1, high);
}

void quicksort(std::vector<int>& nums) { helper(nums, 0, nums.size() - 1); }

int quickselect(std::vector<int>& nums, int low, int high, int k) {
    if (low == high) {
        return nums[low];
    }
    int pivot = partition(nums, low, high);
    if (pivot == k - 1) {
        return nums[pivot];
    } else if (pivot < k - 1) {
        return quickselect(nums, pivot + 1, high, k);
    } else {
        return quickselect(nums, low, pivot - 1, k);
    }
}

int findKthLargest(std::vector<int>& nums, int k) { return quickselect(nums, 0, nums.size() - 1, nums.size() - k + 1); }

int main() {
    {
        std::vector<int> nums = {3, 2, 1, 1, 8, 9, 11};
        quicksort(nums);
        for (auto num : nums) {
            std::cout << num << " ";
        }
        std::cout << '\n';
    }
    {
        std::vector<int> nums = {1, 1, 1};
        quicksort(nums);
        for (auto num : nums) {
            std::cout << num << " ";
        }
        std::cout << '\n';
    }
    {
        std::vector<int> nums = {3, 2, 1};
        quicksort(nums);
        for (auto num : nums) {
            std::cout << num << " ";
        }
        std::cout << '\n';
    }
}

File: Algorithm/quicksort/quicksort/quicksort.go
package main

import "fmt"

func partition(nums []int) int {
	n := len(nums)
	pivot := nums[n-1]
	l := 0
	for r := range n - 1 {
		if nums[r] <= pivot {
			nums[l], nums[r] = nums[r], nums[l]
			l++
		}
	}
	nums[l], nums[n-1] = nums[n-1], nums[l]
	return l
}

func quicksort(nums []int) {
	if len(nums) <= 1 {
		return
	}
	pivot := partition(nums)
	if pivot > 0 {
		quicksort(nums[:pivot])
	}
	if pivot <= len(nums) {
		quicksort(nums[pivot+1:])
	}
}

func findKthLargest(nums []int, k int) int {
	pivot := partition(nums)
	if pivot == k-1 {
		return nums[pivot]
	} else if pivot < k-1 {
		return findKthLargest(nums[pivot+1:], k-pivot-1)
	} else {
		return findKthLargest(nums[:pivot], k)
	}
}

func main() {
	{
		nums := []int{3, 2, 1, 8, 9, 11}
		fmt.Println(findKthLargest(nums, 2))
	}
	{
		nums := []int{1, 2, 3, 4}
		quicksort(nums)
		fmt.Println(nums)
	}
	{
		nums := []int{4, 3, 2, 1}
		quicksort(nums)
		fmt.Println(nums)
	}
	{
		nums := []int{1, 1, 1, 1, 1}
		quicksort(nums)
		fmt.Println(nums)
	}
}
